service/history/historyEngine.go

```diff
@@ -294,6 +302,56 @@ func (e *historyEngineImpl) GetWorkflowExecutionNextEventID(
 		RunId:      request.Execution.RunId,
 	}
 
+	response, err := e.getWorkflowExecutionNextEventID(domainID, execution)
+	if err != nil {
+		return nil, err
+	}
+
+	// expectedNextEventID is 0 when caller want to get the current next event ID without blocking
+	expectedNextEventID := request.GetExpectedNextEventId()
+
+	// if caller decide to long poll on workflow execution
+	// and the event ID we are looking for is smaller than current next event ID
+	if expectedNextEventID >= response.GetEventId() && response.GetIsWorkflowRunning() {
+		subscriberID, channel, err := e.historyEventNotifier.WatchHistoryEvent(&domainID, &execution)
+		if err != nil {
+			return nil, err
+		}
+		defer e.historyEventNotifier.UnwatchHistoryEvent(&domainID, &execution, subscriberID)
+
+		// check again in case the next event ID is updated
+		response, err = e.getWorkflowExecutionNextEventID(domainID, execution)
+		if err != nil {
+			return nil, err
+		}
+
+		if expectedNextEventID < response.GetEventId() || !response.GetIsWorkflowRunning() {
+			return response, nil
+		}
+
+		timeoutChan := time.NewTimer(e.shard.GetConfig().LongPollExpirationInterval).C
```

Comment:
You also need to call timer.Stop(), otherwise you may leak timers.

Reply:
ok